libutils/glob_lib.c

```diff
@@ -0,0 +1,516 @@
+        BufferAppendF(buf, "[%s]", stuff);
+        FREE_AND_NULL(stuff);
+    }
+
+    return left;
+}
+
+/**
+ * Translate a shell pattern to a regular expression.
+ *
+ * This is a C implementation of the translate function in the fnmatch module
+ * from Python's standard library. See
+ * https://github.com/python/cpython/blob/3.8/Lib/fnmatch.py
+ */
+static char *TranslateGlob(const char *const pattern)
+{
+    assert(pattern != NULL);
+
+    static const char *const special_chars = "()[]{}?*+-|^$\\.&~# \t\n\r\v\f";
+    size_t i = 0;
+    const size_t n = strlen(pattern);
+    Buffer *const buf = BufferNew();
+
+    while (i < n)
+    {
+        char ch = pattern[i++];
+
+        switch (ch)
+        {
+#ifndef _WIN32
+        /* We don't allow escaping wildcards in Windows for two reasons. First,
+         * we cannot escape with backslash when it is also a path separator.
+         * Secondly, '*' is an illegal filename on Windows.
+         */
+        case '\\':
+            BufferAppendChar(buf, ch);
+            BufferAppendChar(buf, pattern[i++]);
+            break;
+#endif // _WIN32
+        case '*':
+            BufferAppendString(buf, ".*");
+            break;
+
+        case '?':
+            BufferAppendString(buf, ".");
+            break;
+
+        case '[':
+            i = TranslateBracket(pattern, n, i, buf);
+            break;
+
+        default:
+            if (strchr(special_chars, ch) != NULL)
+            {
+                BufferAppendF(buf, "\\%c", ch);
+            }
+            else
+            {
+                BufferAppendChar(buf, ch);
+            }
+            break;
+        }
+    }
+
+    char *const res = StringFormat("(?s:%s)\\Z", BufferData(buf));
+    BufferDestroy(buf);
+    return res;
+}
+
+bool GlobMatch(const char *const _pattern, const char *const _filename)
+{
+    assert(_pattern != NULL);
+    assert(_filename != NULL);
+
+    char *pattern = NormalizePath(_pattern);
+    char *const filename = NormalizePath(_filename);
+
+    StringSet *const expanded = StringSetNew();
+    ExpandBraces(pattern, expanded);
+    FREE_AND_NULL(pattern);
+
+    bool any_matches = false;
+    StringSetIterator iter = StringSetIteratorInit(expanded);
+    while ((pattern = StringSetIteratorNext(&iter)) != NULL)
+    {
+        char *const regex = TranslateGlob(pattern);
+        any_matches |= StringMatchFull(regex, filename);
+        free(regex);
+    }
+
+    StringSetDestroy(expanded);
+    free(filename);
+
+    return any_matches;
+}
+
+/**
+ * Returns true if basename represents current or parent directory (i.e., '.'
+ * or '..').
+ */
+static inline bool BasenameIsParentOrCurrent(const char *const path)
+{
+    assert(path != NULL);
+
+    char b_copy[strlen(path) + 1];
+    strcpy(b_copy, path);
+    const char *const b_name = basename(b_copy);
+    return StringEqual(b_name, ".") || StringEqual(b_name, "..");
+}
+
+/**
+ * Strategically walk through filesystem looking for files matching pattern.
```

Comment:
I'd replace *Strategically* with *Depth-first* or something more descriptive like that.